### DISCUSSION TITLE
PhysicsStepListener for movement randomly stops working

### DISCUSSION THREAD

**Discussion by hhyyrylainen** [jrouwe/JoltPhysics#694] [Q&A] [answered]
I was debugging an issue where my player character randomly stopped moving, and I think the issue is that `PhysicsStepListener::OnStep` stops being called. From just a quick debugging session (this is the first time I'm debugging Jolt internals), it seems like the body goes to sleep if I don't move the player for a while and then when I'd like to resume moving, that doesn't actually work because no step listeners run on physics updates. Is this an unintended use case for the step listeners? Should I do something else to apply impulses and angular velocity instead? I guess if I always have sub steps kept at one, I can in my physics world wrapper class call the method that the step listener would call.

**Comment by jrouwe** [comment, marked as answer]
Yes, currently when _not a single_ physics body is active in the world, the physics system will early out and do nothing (and also not call the step listeners), this is documented here (and there's even a unit test for it):

https://github.com/jrouwe/JoltPhysics/blob/5d0838e48527a1344e3d3aecea933c043879019f/Jolt/Physics/PhysicsStepListener.h#L23

I'm assuming you use Character instead of CharacterVirtual? For the player character I would recommend using CharacterVirtual which can't sleep because you call the update function yourself. If you don't want to do that you can call ```SetAllowSleeping(false)``` on the Body belonging to the Character, or if you do want the sleeping then you would have to call the listeners yourself when there are no active bodies. For reference, the early out is here:

https://github.com/jrouwe/JoltPhysics/blob/5d0838e48527a1344e3d3aecea933c043879019f/Jolt/Physics/PhysicsSystem.cpp#L132-L134

**Reply by hhyyrylainen** [reply]
Ah, I apparently did not read enough of the comments in the step listener as I missed that. Thanks for pointing that out. And sorry for not thinking first to re-read the documentation on that class.

For my player "character" I'm just actually using just a normal physics body to replicate how it was implemented before with Bullet. I'll test out if the allow sleeping works for my use case or I'll need to manually call the listener.  Is there a way to easily emulate the "lock all bodies" function the step listener has? That way I don't need to switch to locking the bodies one by one to apply the movement control forces.

There's a potential performance penalty if not all "characters" move and all the time and they aren't allowed to sleep, right? So I might want to prefer the manual "step listener" approach.

**Reply by jrouwe** [reply]
> Is there a way to easily emulate the "lock all bodies" function the step listener has?

Well, not a super nice one. You would need to call ```PhysicsSystem::GetBodies``` and then use ```BodyLockMultiWrite``` (which has an optimization when locking a large number of bodies).

> There's a potential performance penalty if not all "characters" move and all the time and they aren't allowed to sleep, right? So I might want to prefer the manual "step listener" approach.

Yes, but you only need 1 active body to keep the simulation alive, so technically speaking you only need to set the flag on a single body (doesn't even need to be the player) and then the overhead will be minimal.

**Reply by hhyyrylainen** [reply]
Thanks for the suggestions. I think I'll just go with a system that makes sure the player's physics body is not allowed to sleep.
